be/src/storage/lake/pk_meta_file.cpp

```diff
@@ -0,0 +1,233 @@
+#include "storage/lake/update_manager.h"
+#include "util/coding.h"
+#include "util/defer_op.h"
+#include "util/json.h"
+#include "util/raw_container.h"
+
+namespace starrocks {
+namespace lake {
+
+PkMetaFileBuilder::PkMetaFileBuilder(const TabletMetadata& metadata) {
+    _tablet_meta = std::make_unique<TabletMetadata>(metadata);
+    _footer = std::make_unique<PKTableFooter>();
+}
+
+void PkMetaFileBuilder::append_delvec(DelVectorPtr delvec, const uint32_t segment_id) {
+    const uint64_t offset = _buf.size();
+    std::string delvec_str;
+    delvec->save_to(delvec_str);
+    _buf.insert(_buf.end(), delvec_str.begin(), delvec_str.end());
+    const uint64_t size = _buf.size() - offset;
+    _delvecs[segment_id].set_segment_id(segment_id);
+    _delvecs[segment_id].mutable_page()->set_offset(offset);
+    _delvecs[segment_id].mutable_page()->set_size(size);
+    TabletSegmentId tsid;
+    tsid.tablet_id = _tablet_meta->id();
+    tsid.segment_id = segment_id;
+    _cache_delvec_updates.push_back(std::make_pair(tsid, delvec));
+}
+
+void PkMetaFileBuilder::apply_opwrite(const TxnLogPB_OpWrite& op_write) {
+    auto rowset = _tablet_meta->add_rowsets();
+    rowset->CopyFrom(op_write.rowset());
+    rowset->set_id(_tablet_meta->next_rowset_id());
+    if (op_write.rowset().num_rows() > 0) {
+        _tablet_meta->set_next_rowset_id(_tablet_meta->next_rowset_id() + rowset->segments_size());
+    } else {
+        // delete
+        _tablet_meta->set_next_rowset_id(_tablet_meta->next_rowset_id() + 1);
+    }
+}
+
+void PkMetaFileBuilder::_finalize_delvec(const int64_t version) {
+    const size_t delvec_size = _tablet_meta->delvec_meta().delvecs_size();
+    for (size_t i = 0; i < delvec_size; i++) {
+        auto* each_delvec = _tablet_meta->mutable_delvec_meta()->mutable_delvecs(i);
+        auto iter = _delvecs.find(each_delvec->segment_id());
+        if (iter != _delvecs.end()) {
+            each_delvec->mutable_page()->set_version(version);
+            each_delvec->mutable_page()->set_offset(iter->second.page().offset());
+            each_delvec->mutable_page()->set_size(iter->second.page().size());
+            _delvecs.erase(iter);
+        }
+    }
+    for (auto&& each_delvec : _delvecs) {
+        each_delvec.second.mutable_page()->set_version(version);
+        _tablet_meta->mutable_delvec_meta()->add_delvecs()->CopyFrom(each_delvec.second);
+    }
+}
+
+Status PkMetaFileBuilder::finalize(LocationProvider* location_provider, UpdateManager* mgr) {
+    const int64_t version = _tablet_meta->version();
+    auto filepath = location_provider->tablet_metadata_location(_tablet_meta->id(), version);
+    auto options = WritableFileOptions{.sync_on_close = true, .mode = FileSystem::CREATE_OR_OPEN_WITH_TRUNCATE};
+    auto writer_file = fs::new_writable_file(options, filepath);
+    if (!writer_file.ok()) {
+        return writer_file.status();
+    }
+    DeferOp close_file([&] { (*writer_file)->close(); });
```

Comment:
Must check the return value of `close()`

Reply:
Done. And above.